### PR TITLE
[SYCL] Support unnamed lambda kernels

### PR DESCRIPTION
**Pull request by bader**
Add support for kernels without kernel name. Requires compiling
with -DUNNAMED_LAMBDA_EXT. Uses new __unique_stable_name
compiler built-in.

### CODE REVIEW

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -352,6 +353,9 @@ class SYCLIntegrationHeader {
     /// Kernel name type.
     QualType NameType;
 
+    /// Kernel name with stable lamba name mangling
```
Comment (mlychkov) [inline clang/include/clang/Sema/Sema.h]:
```suggestion
    /// Kernel name with stable lambda name mangling
```

File: clang/include/clang/Sema/Sema.h [hunk 2]
```diff
@@ -352,6 +353,9 @@ class SYCLIntegrationHeader {
...
+    std::string StableName;
```
Comment (MrSidims) [inline clang/include/clang/Sema/Sema.h]: I suggest to make it StringRef.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 3]
```diff
@@ -9768,7 +9768,7 @@ def err_builtin_launder_invalid_arg : Error<
 // SYCL-specific diagnostics
 def err_sycl_attribute_address_space_invalid : Error<
   "address space is outside the valid range of values">;
-def err_sycl_kernel_name_class_not_top_level : Error<
+def warn_sycl_kernel_name_class_not_top_level : Warning<
```
Comment (Ruyk) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td]: Why does the error gets converted to a warning? *if* the kernel is named, the error still applies...

File: clang/include/clang/Sema/Sema.h (hunk 2 above)
Comment (erichkeane) [reply clang/include/clang/Sema/Sema.h]: This cannot be a stringref.

**Comment by bader** [comment]
> Should this get marked in some way with what it is closing?

Do you mean I should clarify _**why**_ we add "support for unnamed lambda kernels"?

**Comment by erichkeane** [comment]
> 
> 
> > Should this get marked in some way with what it is closing?
> 
> Do you mean I should clarify _**why**_ we add "support for unnamed lambda kernels"?

No, I mean, we have #if <WHATEVER>.  Should the #endif be #endif // <WHATEVER>, or at least have a comment explaining what it is for?  Code-emitter emitting a macro is particularly tough to reason about for the next reader.

**Pull request merged**